submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,241 @@
+    //Specifies the next hop to reach the remote CIDRs
+    NextHop string `json:"nextHop"`
+
+    //Specifies the remote CIDRs available via the next hop
+    RemoteCidr []string `json:"remoteCidr"`
+}
+```
+
+#### SubmarinerNonGWRoute
+
+This CR will be created when a remote endpoint is created. When more remote endpoints are added the Remote CIDR list will be updated.
+
+* NextHop - Specifies the next hop.
+* RemoteCIDR - Specifies the list of remote CIDRs reachable via this gateway.
+
+This CR will be used by the Routeagent pods running on the non-Gateway nodes to
+
+* In non-g/w node - If the route-agent pod is not in the same zone as Gateway node zone, send the traffic to the g/w node zone.
+* In g/w node - add route to send the traffic from other zones, destined to remote cluster to the submariner tunnel
+
+``` go
+type SubmarinerNonGWRoute struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    SubmarinerRoutePolicySpec SubmarinerRoutePolicySpec `json:"submarinerRoutePolicySpec"`
+}
+
+type SubmarinerRoutePolicySpec struct {
+    //Specifies the next hop to reach the remote CIDRs
+    NextHop string `json:"nextHop"`
+
+    //Specifies the remote CIDRs available via the next hop
+    RemoteCidr []string `json:"remoteCidr"`
+}
+```
+
+## Design Details
+
+In this proposal, the plan is to remove the current `submariner-router` and switches that were added to the OVN db. The network-plugin-syncer
+shall be removed  and can be replaced by controllers in and Submariner RouteAgent. Since we have multiple ovn db to
+program, we need multiple connections. So it makes it easier to program the OVN datapath from RouteAgent, than using
+a separate pod. This new approach will work for both IC enabled and existing deployments. With this change we are planning to use the
+ovn-k8s-mp0 interface to reach the host networking stack and then the Submariner tunnel. This interface is used by OVN for host-networking
+traffic with in a cluster and will be present in every node.
+
+![UpdatedTopology](./images/submariner-ovn-ic.png)
+
+### SubmarinerRouteAgentPod
+
+The Submariner Route-agent pod will be responsible for creating the SubmarinerGWRoute CR. It will be used only for OVN CNI right now. For every
+RemoteEndpointCreated event a SubmarinerGWRoute CR will be created. The nextHop will be the interface IP through which we can reach the cable
+driver. In the case of OVN it will be the IP of ovn-k8s-mp0 interface.
+
+The SubmarinerNonGWRoute CRD will also be created by Submariner Route-agent. It will have the list of remote endpoints connected to the gateway.
+The nexthop will be the transit switch IP of the G/W node. If the transit switch IP is missing this CRD will not be created, which means it is
+a non-IC setup.
+
+The Routeagent  will have these controllers added to it and the one running in gateway node responds to the CRUD operations.
+
+#### SubmarinerGWRoute CR Controller
```

Comment:
Do we create the same route [1]  in GW node  for  SubmarinerGWRoute and SubmarinerNonGWRoute CRs ? 


[1] 
bfd                 : []
external_ids        : {}
ip_prefix           : "10.132.0.0/16"
nexthop             : "10.244.1.2"
options             : {}
output_port         : []
policy              : []
route_table         : ""

Reply:
As discussed , a route will be created in ovn-cluster-router of g/w node zone  in response SubmarinerGWRoute and ins response to SubmarinerNonGWRoute a route will be created in ovn-cluster-router of non-g/w zone.